multi_neper.py

```diff
@@ -0,0 +1,253 @@
+
+                for i, dev in enumerate(devices):
+                        if not args.dry_run:
+                                run_pre_neper_cmds(dev)
+
+                        # TODO flow-steering rules installed in Neper now
+                        # control_port = args.control_port + i
+                        # starting_port = i * args.threads + args.source_port
+                        # dev = devices[i]
+                        # src_ip, dst_ip = src_ips[i], hosts[i]
+
+                        # # TODO port_start q_start, q_num
+                        # dst_port = args.port + i
+                        # rules = install_flow_steer_rules(dev, args.threads, starting_port, dst_port, src_ip, dst_ip, args.q_start, args.q_num)
+                        # dev_to_rule[dev] = rules
+
+        cmds = []
+        debug(f"running on {devices}")
+        is_client = args.client
+
+        for i, dev in enumerate(devices):
+                nic_pci = link_to_nic_pci_addr[dev]
+                gpu_pci = link_to_gpu_pci_addr[dev]
+
+                # increment control port by 1, and src/dst ports by flow_count
+                # for each additional link we're running Neper on
+                ctrl_port = args.control_port + i
+                src_port = i * args.flows + args.source_port
+                dst_port = i * args.flows + args.port
+
+                src_ip, dst_ip = src_ips[i], hosts[i]
+
+                # TODO should CPU range be configurable by the user?
+                cpu_range = get_cpu_range(2 + (52 if i >= 2 else 0), args.threads, i)
+
+                cmd_env = build_neper_cmd(args.neper_dir, is_client, dev,
+                                      args.threads, args.flows, cpu_range, args.buffer_size,
+                                      args.phys_len, nic_pci, gpu_pci,
+                                      ctrl_port, src_port, dst_port, args.length, src_ip, dst_ip,
+                                      args.q_start, args.q_num, args.tcpd_validate, args.tcpd_rx_cpy)
+
+                cmds.append(cmd_env)
+
+        for cmd in cmds:
+                debug(cmd)
+
+        if not args.dry_run:
+                sp_list = run_cmds(cmds)
+                debug("parsing subprocesses outputs")
+                for dev, i in zip(devices, parse_subprocess_outputs(sp_list)):
+                        if not args.client:
+                                try:
+                                        print(f"[{dev}] Throughput (Mb/s): {i['throughput']}")
+                                except KeyError:
+                                        print(f"[{dev}] Throughput (Mb/s): NA")
+
+                # TODO remove, flow-steering rules are installed via Neper now
+                # delete flow-steering rules
+                # if not args.client:
+                #         info("deleting flow-steering rules")
+                #         for dev in dev_to_rule:
+                #                 del_flow_steer_rules(dev, dev_to_rule[dev])
```

Comment:
This script looks really great, and makes me appreciate that this was indeed a very complex effort. Thanks!!